src/test/java/org/jenkinsci/plugins/workflow/actions/ErrorActionTest.java

```diff
@@ -228,6 +253,87 @@ public static class X extends Exception {
         });
     }
 
+    @Test public void findOriginFromBodyExecutionCallback() throws Throwable {
+        rr.then(r -> {
+            agents.createAgent(r, "remote");
+            var p = r.createProject(WorkflowJob.class);
+            p.setDefinition(new CpsFlowDefinition("callsFindOrigin {node('remote') {fails()}}", true));
+            var b = p.scheduleBuild2(0).waitForStart();
+            r.waitForMessage("Acting slowly in ", b);
+            agents.stop("remote");
+            r.assertBuildStatus(Result.FAILURE, r.waitForCompletion(b));
+            r.assertLogContains("Found in: fails", b);
+        });
+    }
+    public static final class WrapperStep extends Step {
+        @DataBoundConstructor public WrapperStep() {}
+        @Override public StepExecution start(StepContext context) throws Exception {
+            return new ExecutionImpl(context);
+        }
+        private static final class ExecutionImpl extends StepExecution {
+            ExecutionImpl(StepContext context) {
+                super(context);
+            }
+            @Override public boolean start() throws Exception {
+                getContext().newBodyInvoker().withCallback(new Callback()).start();
+                return false;
+            }
+        }
+        private static class Callback extends BodyExecutionCallback {
+            @Override public void onSuccess(StepContext context, Object result) {
+                context.onSuccess(result);
+            }
+            @Override
+            public void onFailure(StepContext context, Throwable t) {
+                try {
+                    var l = context.get(TaskListener.class);
+                    Functions.printStackTrace(t, l.error("Original failure:"));
+                    l.getLogger().println("Found in: " + ErrorAction.findOrigin(t, context.get(FlowExecution.class)).getDisplayFunctionName());
```

Comment:
Hmm, for this to work, `ErrorAction.equals` would need to be changed to check for ID equality before checking for class equality.

Reply:
And I guess we could do that only if at least one of the exceptions is a `ProxyException` to avoid walking all causes looking for IDs unnecessarily.

Reply:
Indeed

```
FINE	o.j.p.w.actions.ErrorAction#equals: Different types: class hudson.remoting.RequestAbortedException vs. class hudson.remoting.ProxyException
```

Reply:
https://github.com/jenkinsci/workflow-api-plugin/pull/327/commits/c27b80bcdfb3d416e66ae24f2e2cac65b12f6222 might help

Reply:
> avoid walking all causes looking for IDs unnecessarily

This seems like a strange thing to try to optimize, when it is basically just chasing a short chain of pointers.

Reply:
Well, the fallback case when there is no ID is to compare stack trace elements, which can be expensive for big Pipelines when you are looking at all nodes in `ErrorAction.findOrigin`.

Reply:
Right, I meant to swap the order of the checks for id vs. class/message. But this is not enough as seen below.

Reply:
Oh, duh, instead of creating a new `ErrorId` we need to reuse the first one.